test/integration/fake_upstream.cc

```diff
@@ -147,7 +147,18 @@ void FakeConnectionBase::close() {
 }
 
 void FakeConnectionBase::readDisable(bool disable) {
-  connection_.dispatcher().post([this, disable]() -> void { connection_.readDisable(disable); });
+  std::unique_lock<std::mutex> lock(lock_);
+  if (!disconnected_) {
+    connection_.dispatcher().post([this, disable]() -> void {
+      if (!disconnected_) {
+        connection_.readDisable(disable);
+      }
```

Comment:
Should we put an else clause here with NOT_IMPLEMENTED? Seems like race could still happen? Or can that one happen for real?